### PR TITLE
Deactivate the tracking of gradients in sampling functions within OpInfos

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67833
* #68188
* #67996
* #68523
* __->__ #68522
* #69909
* #69908

Some OpInfos were inadvertibly generating samples with `grad_fn`. For
example, when using functions like `transpose()` or `conj()` on the
inputs to generate transposed or conjugated inputs. This PR corrects
this and deactivates the tracking of gradients in all the sampling
functions.

Differential Revision: [D33774420](https://our.internmc.facebook.com/intern/diff/D33774420)

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -655,7 +655,7 @@ def __init__(self,
 
         self.decorators = (*decorators, *skips)
 
-        self.sample_inputs_func = sample_inputs_func
+        self.sample_inputs_func = torch.no_grad()(sample_inputs_func)
```
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:658]: nb. This is the important point of this PR. The rest of the PR just rewrites some tests for them to work as expected with this change and with the `non_contiguous` tests added in https://github.com/pytorch/pytorch/pull/67996

**Review by anjali411 — APPROVED** [review]
LGTM!

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -1946,8 +1941,7 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad, **kwargs):
         sample_inputs.append(
             SampleInput(make_tensor(shape, device, dtype, requires_grad=requires_grad),
                         args=(
-                            make_tensor(shape, device, dtype,
-                                        requires_grad=requires_grad).t().conj(),
+                            make_tensor(shape, device, dtype).mH.requires_grad_(requires_grad),
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:1944]: do you get an error if you don't set the requires grad? I know you left a note above to set requires grad after performing any transformations but I am worried because it's an easy thing to miss

Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:1944]: Yes, you do get an error. I needed to update all these functions for the tests to pass.

Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py:1944]: maybe we should create helper functions for standard transformations that people can use. We can take care of retaining the `requires_grad` field in those helper functions then!

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -325,6 +325,10 @@ def _getattr_qual(obj, name, default=_NOTHING):
 #   Each SampleInput defines an "input", "args", "kwargs",
 #   an "output_process_fn_grad" function, the "broadcasts_input" bool and
 #   a "name".
+#   All the "sample_inputs" functions are invoked within a `torch.no_grad()`
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:328]: Newline (new paragraph)

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:662]: Comment here too

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -1894,26 +1895,20 @@ def sample_inputs_t(op_info, device, dtype, requires_grad, **kwargs):
 
 
 def sample_inputs_mm(op_info, device, dtype, requires_grad, **kwargs):
-    make_arg = partial(make_tensor, device=device, dtype=dtype)
+    make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
 
-    first_shape, second_shape = (S, M), (M, S)
-    sample_inputs = []
+    def make_arg_conj(size):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1900]: Wait what does this function do? Doesn't it need a conj()?

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:1900]: Follow-up question for @anjali411 -- should sample inputs be generating conjugated samples, or are those automatically generated now?

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -325,6 +325,10 @@ def _getattr_qual(obj, name, default=_NOTHING):
 #   Each SampleInput defines an "input", "args", "kwargs",
 #   an "output_process_fn_grad" function, the "broadcasts_input" bool and
 #   a "name".
+#   All the "sample_inputs" functions are invoked within a `torch.no_grad()`
+#   environment for efficiency and correctness. As such remember to set the the
+#   "requires_grad" flag on the inputs **after** performing any transformations
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:330]: What's a little scary about this change is that we have no way of validating if the things that should require grad still do after this, right? Like that evaluation is just done manually?

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -2117,9 +2105,9 @@ def sample_inputs_baddbmm(op_info, device, dtype, requires_grad, **kwargs):
                             requires_grad=requires_grad))
         sample_inputs.append(
             SampleInput(
-                args[0].transpose(-1, 1).detach().requires_grad_(requires_grad),
-                args=(args[1].transpose(-1, 1).conj().detach().requires_grad_(requires_grad),
-                      args[2].transpose(-1, 1).conj().detach().requires_grad_(requires_grad)),
+                args[0].transpose_(-1, 1),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2108]: No requires grad here?

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -5263,29 +5233,19 @@ def c(t):
 
 
 def sample_inputs_linalg_qr(op_info, device, dtype, requires_grad=False, **kwargs):
-    """
-    This function generates input for torch.linalg.qr
-    The input is generated as the itertools.product of 'batches' and 'ns'.
-    """
+    # QR is just well defined when the matrix is full rank
+    make_fullrank = make_fullrank_matrices_with_distinct_singular_values
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5237]: haha, that's one way to shorten line length

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -5263,29 +5233,19 @@ def c(t):
 
 
 def sample_inputs_linalg_qr(op_info, device, dtype, requires_grad=False, **kwargs):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5235]: Add a comment that this is also used for geqrf

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -5412,10 +5372,9 @@ def make_arg(shape):
     def prod_zeros(dim_select):
         assert len(dim_select) == 2
         result = make_arg(3 * (S,))
-        with torch.no_grad():
-            result.narrow(dim_select[0], 0, 1).narrow(dim_select[1], 1, 1).zero_()
-            result.narrow(dim_select[0], 2, 1).narrow(dim_select[1], 3, 1).zero_()
-            result.narrow(dim_select[0], 4, 1).narrow(dim_select[1], 3, 1).zero_()
+        result.narrow(dim_select[0], 0, 1).narrow(dim_select[1], 1, 1).zero_()
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5375]: The result of `narrow()` on a tensor that requires grad is a tensor that requires grad even under the no grad context?

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -6017,103 +5974,50 @@ def sample_inputs_i0_i1(op_info, device, dtype, requires_grad, **kwargs):
...
-        filtered_product = filter(filter_fn, product(args, alphas))  # type: ignore[var-annotated]
-        return (SampleInput(input.detach().clone().requires_grad_(requires_grad),
-                            args=(arg,), kwargs=dict(alpha=alpha))
-                for (input, arg), alpha in filtered_product)
+def sample_inputs_rsub(op_info, device, dtype, requires_grad, other_scalar, **kwargs):
+    make_arg = partial(make_tensor, device=device)
 
-    int_alpha, float_alpha, complex_alpha = 2, 0.1, 1 + 0.6j
+    shapes = ((S, S), (S,), ())
+    if dtype.is_complex:
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5996]: This needs a comment

File: torch/testing/_internal/common_methods_invocations.py [hunk 11]
```diff
@@ -6017,103 +5974,50 @@ def sample_inputs_i0_i1(op_info, device, dtype, requires_grad, **kwargs):
...
-def sample_inputs_rsub(op_info, device, dtype, requires_grad, variant='tensor', **kwargs):
-    def _make_tensor_helper(shape, low=None, high=None):
-        return make_tensor(shape, device, dtype, low=low, high=high, requires_grad=requires_grad)
-
-    def _samples_with_alpha_helper(args, alphas, filter_fn=lambda arg_alpha: True):
-        filtered_product = filter(filter_fn, product(args, alphas))  # type: ignore[var-annotated]
-        return (SampleInput(input.detach().clone().requires_grad_(requires_grad),
-                            args=(arg,), kwargs=dict(alpha=alpha))
-                for (input, arg), alpha in filtered_product)
+def sample_inputs_rsub(op_info, device, dtype, requires_grad, other_scalar, **kwargs):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5992]: Don't we have generic elementwise binary sample inputs now for ops like add and sub? Can this just reuse that function?

File: torch/testing/_internal/common_methods_invocations.py [hunk 12]
```diff
@@ -13002,8 +12902,6 @@ def ref_pairwise_distance(input1, input2):
             DecorateInfo(unittest.skip("Skipped!"), 'TestReductions', 'test_dim_none'),
             DecorateInfo(unittest.skip("Skipped!"), 'TestReductions', 'test_dim_none_keepdim'),
             # FIXME: improve precision
-            DecorateInfo(unittest.skip("Skipped!"), 'TestReductions', 'test_noncontiguous_all',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:13005]: Interesting... what fixed this?

**Review by mruberry — COMMENTED** [review]
Another super cool and interesting PR, @Lezcano. cc @saketh-are @pmeier @kshitij12345 and @krshrimali.

A few questions/comments inline for your review

File: torch/testing/_internal/common_methods_invocations.py (hunk 9 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:5375]: The result of narrow is a new tensor which is a view on a part of a tensor, that then is zeroed out in place. As such, the original tensor object is not modified (although its values are)

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:5237]: I should really put up a PR renaming this function...

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:2108]: No, as this just modifies the strides of the tensor, but does not return a new tensor.

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:330]: It is true that, if no input from an opinfo has `requires_grad=True`, then we throw, but it could be that there are some operations that return just one of the inputs with `requries_grad` but not all of them.

One way to check this would be to throw if for an input of floating point, none of the inputs has `requires_grad=True`. This might give a few false positives, but I think it might even uncover some previous errors we missed.

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:1900]: Oh, sorry, I completely missed the `conj()`! Thanks for the catch.

Once we test conjugated inputs on all the possible inputs (similar to the extensions we did for `noncontiguous`), these samples will not be necessary.

File: torch/testing/_internal/common_methods_invocations.py (hunk 11 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:5992]: We do have `sample_inputs_rbinops`. Now, it seems rather simple. I wonder whether we should use a more involved scheme like hte one proposed here that tests for tensor-tensor / scalar-tensor operations with different dtypes and all. What do you think?

File: torch/testing/_internal/common_methods_invocations.py (hunk 12 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:13005]: I believe it was failing due to some problems in the previous generation of the inputs (perhaps some flags of the final tensor were not set properly). After rewriting the sampling I run all these to see if it had fixed any of the problems and...

**Comment by Lezcano** [comment]
I just added the suggested changes / answered some of the questions @mruberry

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py:2031]: @mruberry they are generated automatically for all relevant tests I think. 

I added this logic manually to generate conjugated samples when required:

```
include_conjugated_inputs = op.test_conjugated_samples and dtype.is_complex
        samples = op.sample_inputs(device, dtype, requires_grad=_requires_grad, include_conjugated_inputs=include_conjugated_inputs)
```

I see that we are not generating conjugated samples in `test_noncontiguous_samples`!

File: torch/testing/_internal/common_methods_invocations.py [hunk 13]
```diff
@@ -5297,27 +5283,18 @@ def sample_inputs_lu_solve(op_info, device, dtype, requires_grad=False, **kwargs
             b_.requires_grad_(b_grad)
             yield SampleInput(b_, args=(lu_, pivs))
 
+            lu_ = lu.detach().clone()
+            lu_.requires_grad_(lu_grad)
+            b_ = b.detach().clone()
+            b_.requires_grad_(b_grad)
+            yield SampleInput(b_, args=(lu_, pivs))
+
```
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:5291]: remove

**Review by mruberry — APPROVED** [review]
Cool! Just one duplicate code block

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774420).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774420).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774420).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774420).
